src/patterns/ec2-app.test.ts

```diff
@@ -0,0 +1,156 @@
+import "@aws-cdk/assert/jest";
+import { SynthUtils } from "@aws-cdk/assert";
+import { alphabeticalTags, simpleGuStackForTesting } from "../utils/test";
+import { GuApplicationPorts, GuEc2App, GuNodeApp, GuPlayApp } from "./ec2-app";
+
+describe("the GuEC2App pattern", function () {
+  it("should produce a functional EC2 app with minimal arguments", function () {
+    const stack = simpleGuStackForTesting();
+    new GuEc2App(stack, {
+      applicationPort: GuApplicationPorts.Node,
+      app: "test-gu-ec2-app",
+      publicFacing: false,
+      userData: "#!/bin/dev foobarbaz",
+    });
+    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
+  });
+
+  it("can handle multiple EC2 apps in a single stack", function () {
+    const stack = simpleGuStackForTesting();
+    new GuEc2App(stack, {
+      applicationPort: GuApplicationPorts.Node,
+      app: "NodeApp",
+      publicFacing: false,
+      userData: "#!/bin/dev foobarbaz",
+    });
+
+    new GuEc2App(stack, {
+      applicationPort: GuApplicationPorts.Play,
+      app: "PlayApp",
+      publicFacing: false,
+      userData: "#!/bin/dev foobarbaz",
+    });
+
+    expect(stack).toHaveResource("AWS::AutoScaling::AutoScalingGroup", {
+      Tags: alphabeticalTags([
+        {
+          Key: "App",
+          PropagateAtLaunch: true,
+          Value: "PlayApp",
+        },
+        {
+          Key: "gu:cdk:version",
+          PropagateAtLaunch: true,
+          Value: "TEST",
+        },
```

Comment:
Better to use the [mock](https://github.com/guardian/cdk/blob/d69bdbe181d90f0f2c502dee9b05095f1402301a/src/constants/__mocks__/library-info.ts) here.

This would become:

```typescript
import { TrackingTag } from "../../constants/library-info";

...

Tags: alphabeticalTags([
  TrackingTag,
  { Key: "App", ... },
  ...
]);
```

That is, this can be replaced:

```typescript
{
  Key: "gu:cdk:version",
  PropagateAtLaunch: true,
  Value: "TEST",
},
```

With:

```typescirpt
TrackingTag
```

Using the constant means we can update the naming of the tag in one place (should we want to).

Reply:
Good shout. This doesn't currently work due to the missing `PropagateAtLaunch` keypair, but I've created a new constant that handles this called `TrackingTagWithPropagate`.